File: crm/app/internal/domain/conversation.go
// Package domain — Conversation and Message entities.
// A Conversation is a thread of messages on an external channel (Messenger, Zalo, Shopee).
// No adapter imports allowed here.
package domain

import "fmt"

// Conversation tracks a chat thread between a customer and the shop.
// party_id is nullable until the external sender (psid) is resolved to a CRM party.
type Conversation struct {
	ConversationID   string  // UUID
	PartyID          *string // FK → crm_party (nullable)
	Channel          string  // messenger|shopee|zalo
	ExternalThreadID string  // psid (Messenger) / buyer id (Shopee) / user id (Zalo)
	PageID           string  // FB page / Shopee shop / Zalo OA identifier
	Status           string  // open|pending|closed
	AssigneeUserID   *string // FK → crm_app_user nullable
	LastMessageAt    *string // UTC ISO-8601 nullable
	UnreadCount      int
	CreatedAt        string // UTC ISO-8601
	UpdatedAt        string // UTC ISO-8601
}

// Message is one inbound or outbound message within a conversation.
type Message struct {
	MessageID         string  // UUID
	ConversationID    string  // FK → crm_conversation
	ExternalMessageID string  // channel-native message id (idempotency key)
	Direction         string  // in|out
	SenderRef         string  // psid, page_id, or staff identifier
	Body              string
	Attachments       string  // JSON array (TEXT column); empty string = no attachments
	SentAt            string  // UTC ISO-8601
}

// ValidConversationStatuses lists accepted status values.
var ValidConversationStatuses = []string{"open", "pending", "closed"}

// IsValidConversationStatus returns true when s is an accepted conversation status.
func IsValidConversationStatus(s string) bool {
	for _, v := range ValidConversationStatuses {
		if v == s {
			return true
		}
	}
	return false
}

// allowedConvTransitions maps current status → reachable next statuses.
var allowedConvTransitions = map[string][]string{
	"open":    {"pending", "closed"},
	"pending": {"open", "closed"},
	"closed":  {"open"},
}

// CanTransitionConvTo returns nil when current → next is a valid conversation transition.
func CanTransitionConvTo(current, next string) error {
	if current == next {
		return nil
	}
	targets, ok := allowedConvTransitions[current]
	if !ok {
		return fmt.Errorf("conversation: unknown current status %q", current)
	}
	for _, t := range targets {
		if t == next {
			return nil
		}
	}
	return fmt.Errorf("conversation: transition %q → %q is not allowed", current, next)
}

// ParsedMessage is the normalised representation produced by a chat-source adapter
// (e.g. MessengerAdapter) before it is passed to ConversationService.IngestMessage.
// This is the seam between inbound adapters and the application layer.
type ParsedMessage struct {
	Channel           string // messenger|shopee|zalo
	ExternalThreadID  string // psid (sender id for Messenger)
	PageID            string // FB page / shop / OA id
	ExternalMessageID string // channel-native message id
	Direction         string // in|out
	SenderRef         string // who sent (psid or page_id)
	Body              string
	Attachments       string  // JSON array or empty string
	SentAt            string  // UTC ISO-8601
}

File: crm/python/domain/entities/conversation.py
"""CRM Conversation and Message domain entities — chat threads on external channels.

Pure dataclasses; no HTTP/DB adapter imports.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

# ---------------------------------------------------------------------------
# Channel constants
# ---------------------------------------------------------------------------
CHANNEL_MESSENGER = "messenger"
CHANNEL_SHOPEE = "shopee"
CHANNEL_ZALO = "zalo"
VALID_CHANNELS = [CHANNEL_MESSENGER, CHANNEL_SHOPEE, CHANNEL_ZALO]

# ---------------------------------------------------------------------------
# Conversation status constants
# ---------------------------------------------------------------------------
CONV_STATUS_OPEN = "open"
CONV_STATUS_PENDING = "pending"
CONV_STATUS_CLOSED = "closed"
VALID_CONVERSATION_STATUSES = [CONV_STATUS_OPEN, CONV_STATUS_PENDING, CONV_STATUS_CLOSED]

# ---------------------------------------------------------------------------
# Allowed conversation status transitions
# ---------------------------------------------------------------------------
CONV_ALLOWED_TRANSITIONS: dict[str, list[str]] = {
    CONV_STATUS_OPEN: [CONV_STATUS_PENDING, CONV_STATUS_CLOSED],
    CONV_STATUS_PENDING: [CONV_STATUS_OPEN, CONV_STATUS_CLOSED],
    CONV_STATUS_CLOSED: [CONV_STATUS_OPEN],
}

# ---------------------------------------------------------------------------
# Message direction constants
# ---------------------------------------------------------------------------
DIRECTION_IN = "in"
DIRECTION_OUT = "out"


# ---------------------------------------------------------------------------
# Entities
# ---------------------------------------------------------------------------

@dataclass
class Conversation:
    """Chat thread between a customer and the shop on an external channel.

    party_id is nullable until the external sender (psid) is resolved to a CRM party.
    """
    conversation_id: str        # UUID
    channel: str                # messenger|shopee|zalo
    external_thread_id: str     # psid (Messenger) / buyer id (Shopee) / user id (Zalo)
    status: str                 # open|pending|closed
    unread_count: int
    created_at: str             # UTC ISO-8601
    updated_at: str             # UTC ISO-8601
    party_id: Optional[str] = None          # FK → crm_party (nullable)
    page_id: Optional[str] = None           # FB page / Shopee shop / Zalo OA identifier
    assignee_user_id: Optional[str] = None  # FK → crm_app_user nullable
    last_message_at: Optional[str] = None   # UTC ISO-8601 nullable


@dataclass
class Message:
    """One inbound or outbound message within a conversation."""
    message_id: str             # UUID
    conversation_id: str        # FK → crm_conversation
    direction: str              # in|out
    sent_at: str                # UTC ISO-8601
    created_at: str             # UTC ISO-8601
    external_message_id: Optional[str] = None  # channel-native message id (idempotency key)
    sender_ref: Optional[str] = None           # psid, page_id, or staff identifier
    body: Optional[str] = None
    attachments: list = field(default_factory=list)  # parsed from JSON array column


@dataclass
class ParsedMessage:
    """Normalised inbound message produced by a chat-source adapter.

    Seam between inbound adapters and the application layer; passed to
    ConversationService.ingest_message before persistence.
    """
    channel: str                # messenger|shopee|zalo
    external_thread_id: str     # psid (sender id for Messenger)
    page_id: str                # FB page / shop / OA id
    external_message_id: str    # channel-native message id
    direction: str              # in|out
    sender_ref: str             # who sent (psid or page_id)
    sent_at: str                # UTC ISO-8601
    body: str = ""
    attachments: str = ""       # JSON array or empty string (raw column value)
